### ISSUE TITLE
Overzoom Artifacts

### ISSUE THREAD

**Issue by mikeomeara1** [mapbox/mapbox-gl-js#7754]
After generating a number of vector tile sets with Tippecanoe, we've started to notice what appears to be an overzoom rendering issue. It looks like this issue was reported previously in [#6097](https://github.com/mapbox/mapbox-gl-js/issues/6097) but seemed to die on the vine a couple days later (which I'm guessing is due to the lack of an example?).

These artifacts behave exactly as described in [#6097](https://github.com/mapbox/mapbox-gl-js/issues/6097). They appear generally between z16 and z22 and seem to "fix themselves" 2-5 higher zoom levels from when they appeared.

Also as described in the other issue, this appears on all the tilesets we've created spanning a wide variety of geometry density, size and complexity....although it seems to impact smaller/denser geometries more heavily. 

I've also confirmed that this issue exists on tilesets generated directly from GeoJSON in Mapbox Studio (so it doesn't appear to be a Tippecanoe issue).

**mapbox-gl-js version**:
v0.50.0
v0.52.0

**browser**:
Confirmed issue in:
Chrome
Firefox
Edge

### Steps to Trigger Behavior
See Demo for an example

### Link to Demonstration
[https://jsfiddle.net/mikeo1/5xfsdL97/6/](https://jsfiddle.net/mikeo1/5xfsdL97/6/)

- Add an access token
- Run example
- You'll see an example of this behavior
- Zoom in one more tick, issue disappears

For non-Tippecanoe tileset example. See:
Tileset: mikeomeara1.ba0onr77
Zoom: 20.867
Lat: 33.742058534525654
Lng: -117.83724305726696

Please let me know if I can provide any further detail.

Thanks!

**Comment by mikeomeara1** [comment]
![mapbox-overzoom-issue-z19](https://user-images.githubusercontent.com/8380449/50859973-9b445d80-1349-11e9-9928-fc35b6f6ecc4.png)
![mapbox-overzoom-issue-z20](https://user-images.githubusercontent.com/8380449/50859975-9d0e2100-1349-11e9-820a-e7ce5dbb60bf.png)

**Comment by mourner** [comment]
I'm seeing that the source's maxzoom is 11 here. When you zoom to much higher zoom levels, there is simply not enough precision to render the lines correctly, so the gaps may appear — currently we have a technical limit of 8192 extent when rendering tiles. Meaning that 5 zooms higher, it will snap to a 256x256 grid, which is too low res. Do you use the `--detect-shared-borders` option in `tippecanoe`? It might help somewhat.

Let's close this as a duplicate of #6957.

**Comment by mikeomeara1** [comment]
Thanks for the feedback @mourner. Yes, we're still seeing this occur when using `--detect-shared-borders` (but as you've noted it does seem to help). It sounds like we just have to raise our `tippecanoe` `-z` level to something like 15 or 16 (or higher) to make this work. 

Appreciate the clarification!

**Comment by mikeomeara1** [comment]
In case it helps anyone else out there...we found that a little bit of creative styling can help work around this situation in some cases.

We found that `line` layers are much more susceptible to this issue than `fill` layers with a `fill-outline-color`. 

What we did was take the max zoom of your tileset (in my case z11), add 5 (z16 - which as @mourner points out - is where this issue starts to occur) and set that as our `maxzoom` on our `line` layer. This makes it disappear and the `fill-outline-color` of the `fill` layer takes over. 

You'll never be able to have a `line-weight` of more than what `fill-outline-color` gives you, but it does seem to clear up the problem without having to re-gen tilesets at a forced higher zoom level. And I can live with that.

This causes the layer to go from:
![no-max-zoom](https://user-images.githubusercontent.com/8380449/50989298-d1a5e800-14c3-11e9-915e-bec6d8bc94b0.png)

to:
![max-zoom](https://user-images.githubusercontent.com/8380449/50989306-d7033280-14c3-11e9-8ddc-7dd38734279e.png)

Updated fiddle here: [https://jsfiddle.net/mikeo1/5xfsdL97/14/](https://jsfiddle.net/mikeo1/5xfsdL97/14/)
